sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py

```diff
@@ -0,0 +1,287 @@
+
+    # Check that training set can be recovered exactly:
+    predictions = predictor.predict_binned(X_binned)
+    assert_array_almost_equal(predictions, -all_gradients, decimal=5)
+
+
+@pytest.mark.parametrize(
+    'n_samples, min_samples_leaf, n_bins, constant_hessian, noise',
+    [
+        (11, 10, 7, True, 0),
+        (13, 10, 42, False, 0),
+        (56, 10, 255, True, 0.1),
+        (101, 3, 7, True, 0),
+        (200, 42, 42, False, 0),
+        (300, 55, 255, True, 0.1),
+        (300, 301, 255, True, 0.1),
+    ]
+)
+def test_min_samples_leaf(n_samples, min_samples_leaf, n_bins,
+                          constant_hessian, noise):
+    rng = np.random.RandomState(seed=0)
+    # data = linear target, 3 features, 1 irrelevant.
+    X = rng.normal(size=(n_samples, 3))
+    y = X[:, 0] - X[:, 1]
+    if noise:
+        y_scale = y.std()
+        y += rng.normal(scale=noise, size=n_samples) * y_scale
+    mapper = _BinMapper(max_bins=n_bins)
+    X = mapper.fit_transform(X)
+
+    all_gradients = y.astype(G_H_DTYPE)
+    if constant_hessian:
+        all_hessians = np.ones(shape=1, dtype=G_H_DTYPE)
+    else:
+        all_hessians = np.ones_like(all_gradients)
+    grower = TreeGrower(X, all_gradients, all_hessians,
+                        max_bins=n_bins, shrinkage=1.,
+                        min_samples_leaf=min_samples_leaf,
+                        max_leaf_nodes=n_samples)
+    grower.grow()
+    predictor = grower.make_predictor(
+        bin_thresholds=mapper.bin_thresholds_)
+
+    if n_samples >= min_samples_leaf:
+        for node in predictor.nodes:
+            if node['is_leaf']:
+                assert node['count'] >= min_samples_leaf
+    else:
+        assert predictor.nodes.shape[0] == 1
+        assert predictor.nodes[0]['is_leaf']
+        assert predictor.nodes[0]['count'] == n_samples
+
+
+@pytest.mark.parametrize('n_samples, min_samples_leaf', [
+                         (99, 50),
+                         (100, 50)])
+def test_min_samples_leaf_root(n_samples, min_samples_leaf):
+    # Make sure root node isn't split if n_samples is not at least twice
+    # min_samples_leaf
+    rng = np.random.RandomState(seed=0)
+
+    max_bins = 255
+
+    # data = linear target, 3 features, 1 irrelevant.
+    X = rng.normal(size=(n_samples, 3))
+    y = X[:, 0] - X[:, 1]
+    mapper = _BinMapper(max_bins=max_bins)
+    X = mapper.fit_transform(X)
+
+    all_gradients = y.astype(G_H_DTYPE)
+    all_hessians = np.ones(shape=1, dtype=G_H_DTYPE)
+    grower = TreeGrower(X, all_gradients, all_hessians,
+                        max_bins=max_bins, shrinkage=1.,
+                        min_samples_leaf=min_samples_leaf,
+                        max_leaf_nodes=n_samples)
+    grower.grow()
+    if n_samples >= min_samples_leaf * 2:
+        assert len(grower.finalized_leaves) >= 2
+    else:
+        assert len(grower.finalized_leaves) == 1
+
+
+def test_init_parameters_validation():
```

Comment:
I would split the test into 2 tests (input validation and parameters validation):

```python
@pytest.mark.parametrize(
    "X, err_msg, err_type",
    [(_make_training_data()[0].astype(np.float32), 'feature binning required',
      NotImplementedError),
     (np.ascontiguousarray(_make_training_data()[0]), 'as Fortran contiguous',
      ValueError)]
)
def test_input_validation(X, err_msg, err_type):
    _, all_gradients, all_hessians = _make_training_data()
    with pytest.raises(err_type, match=err_msg):
        TreeGrower(X, all_gradients, all_hessians)


@pytest.mark.parametrize(
    'params, err_msg',
    [({'min_gain_to_split': -1}, 'min_gain_to_split=-1 must be positive'),
     ({'min_hessian_to_split': -1}, 'min_hessian_to_split=-1 must be positiv')]
)
def test_init_parameters_validation(params, err_msg):
    X_binned, all_gradients, all_hessians = _make_training_data()
    with pytest.raises(ValueError, match=err_msg):
        TreeGrower(X_binned, all_gradients, all_hessians, **params)
```

Reply:
done, but didn't parametrize since it doesn't help readability here